src/axom/core/Array.hpp

```diff
@@ -55,11 +88,152 @@ bool operator==(const Array<T>& lhs, const Array<T>& rhs);
+   *
+   * \note equivalent to *(array.data() + idx).
+   *
+   * \pre sizeof...(Args) == DIM
+   * \pre 0 <= args[i] < m_dims[i] for i in [0, DIM)
+   */
+  template <typename... Args,
+            typename SFINAE = typename std::enable_if<sizeof...(Args) == DIM>::type>
+  T& operator()(Args... args)
+  {
+    IndexType indices[] = {static_cast<IndexType>(args)...};
+    IndexType idx =
+      std::inner_product(indices, indices + DIM, m_strides.begin(), 0);
+    // assert(inBounds(idx));
+    return asDerived().data()[idx];
+  }
+  /// \overload
+  template <typename... Args,
+            typename SFINAE = typename std::enable_if<sizeof...(Args) == DIM>::type>
+  const T& operator()(Args... args) const
+  {
+    IndexType indices[] = {static_cast<IndexType>(args)...};
+    IndexType idx =
+      std::inner_product(indices, indices + DIM, m_strides.begin(), 0);
+    // assert(inBounds(idx));
+    return asDerived().data()[idx];
+  }
+
+  /// \brief Swaps two ArrayImpls
+  friend void swap(ArrayImpl& lhs, ArrayImpl& rhs)
+  {
+    std::swap(lhs.m_dims, rhs.m_dims);
+    std::swap(lhs.m_strides, rhs.m_strides);
+  }
+
+  /// \brief Returns the dimensions of the Array
+  const std::array<IndexType, DIM>& shape() const { return m_dims; }
+
+  /// \brief Returns the strides of the Array
+  const std::array<IndexType, DIM>& strides() const { return m_strides; }
+
+private:
+  ArrayType& asDerived() { return static_cast<ArrayType&>(*this); }
+  const ArrayType& asDerived() const
+  {
+    return static_cast<const ArrayType&>(*this);
+  }
+
+  // Is it sufficient to just give them the raw arrays?
+  /// \brief The sizes (extents?) in each dimension
+  std::array<IndexType, DIM> m_dims;
+  /// \brief The strides in each dimension
+  std::array<IndexType, DIM> m_strides;
+};
+
+/// \brief Array implementation specific to 1D Arrays
+template <typename T>
+class ArrayImpl<T, 1>
+{
+public:
+  using ArrayType = Array<T, 1>;
+  ArrayImpl(IndexType = 0) { }
+
+  /*!
+   * \brief Push a value to the back of the array.
+   *
+   * \param [in] value the value to the back.
+   *
+   * \note Reallocation is done if the new size will exceed the capacity.
+   */
+  void push_back(const T& value);
+
+  /*!
+   * \brief Push a value to the back of the array.
+   *
+   * \param [in] value the value to move to the back.
+   *
+   * \note Reallocation is done if the new size will exceed the capacity.
+   */
+  void push_back(T&& value);
+
+  /*!
+   * \brief Inserts new element at the end of the Array.
+   *
+   * \param [in] args the arguments to forward to constructor of the element.
+   *
+   * \note Reallocation is done if the new size will exceed the capacity.
+   * \note The size increases by 1.
+   */
+  template <typename... Args>
+  void emplace_back(Args&&... args);
+
+  /// \brief Returns the dimensions of the Array
+  // FIXME: std::array is used for consistency with multidim case, should we just return the scalar?
```

Comment:
Should there be a flag to toggle based on is it multi-dimensional?

Something like:
```
 if (myArray.is_multidimensonal() && myArray.size() == 10)
```

Reply:
or more generically:

```
if (myArray.dimension() == 1 && myArray.size() == 10)
```

Reply:
@kennyweiss  do you have thoughts on this?  The alternate, as is behavior, is to always take the first in the tuple for the 1D case:

```
if (myArray.size()[0] == 10
```

I don't feel strongly either way...

Reply:
Returning an array of size 1 in the 1D case is consistent with `ndarray.shape()` - since the dimension is a template parameter you'd always know if you're dealing with a 1D array anyway, in which case you'd just use `arr.size()` instead of `arr.shape()[0]`.  I could really only imagine calling `shape` on a 1D array in dimension-generic code in which case you'd probably want the return type to be the same regardless of dimension

Reply:
That makes sense and I am okay with moving forward with this as is.